syntax = "proto3";

package services;

service FileService {
    rpc UploadFile (UploadFileRequest) returns (UploadFileResponse);
    rpc DeleteFile (DeleteFileRequest) returns (DeleteFileResponse);
    rpc RenameFile (RenameFileRequest) returns (RenameFileResponse);
}

message UploadFileRequest {
    string filename = 1;
    bytes content = 2;
}

message UploadFileResponse {
    bool success = 1;
}

message DeleteFileRequest {
    string filename = 1;
}

message DeleteFileResponse {
    bool success = 1;
}

message RenameFileRequest {
    string old_filename = 1;
    string new_filename = 2;
}

message RenameFileResponse {
    bool success = 1;
}

service ComputationService {
    rpc Add (AddRequest) returns (AddResponse);
    rpc Sort (SortRequest) returns (SortResponse);

    rpc AddAsync (AddRequest) returns (AckResponse);
    rpc SortAsync (SortRequest) returns (AckResponse);
    rpc GetResult (ResultRequest) returns (ResultResponse);
}

message AddRequest {
    int32 i = 1;
    int32 j = 2;
}

message AddResponse {
    int32 result = 1;
}

message SortRequest {
    repeated int32 array = 1;
}

message SortResponse {
    repeated int32 sorted_array = 1;
}

message AckResponse {
    string ack_id = 1;
}

message ResultRequest {
    string ack_id = 1;
}

message ResultResponse {
    oneof result {
        int32 add_result = 1;
        SortResponse sort_result = 2;
    }
}
